nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,169 @@
+
+
+async def get_signed_cache_value(pool: AIORedis, secret: str, cache_key: str):
+    pipeline = pool.pipeline()
+    pipeline.get(cache_key).get(f"{cache_key}-checksum")
+    pickled_value, cache_checksum = await pipeline.execute()
+    if not pickled_value or not cache_checksum:
+        return None
+
+    cache_checksum = cache_checksum.decode("utf-8")
+    if (checksum := get_hmac_checksum(secret, pickled_value)) != cache_checksum:
+        logger.error(
+            "Checksum for cache was wrong, someone tampered with the values!",
+            correct_checksum=str(cache_checksum),
+            recalculated_checksum=checksum,
+        )
+        return None
+
+    return pickle.loads(pickled_value)
+
+
+def cached_result(pool: AIORedis, prefix: str, secret: str, key_name: str | None, expiry: int = 120) -> Callable:
+    """Pass returned result objects from a function call into redis.
+
+    Returns a decorator function that will cache every result of a function to redis. This only works
+    for functions with string, int or UUID arguments and result objects that can be serialized by the
+    python pickle library.
+
+    Example:
+        Todo: switch to cPickle?
+        Decorate a suitable function with all the default settings like this:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM")
+            def my_cached_function(uuid_arg, string_arg, int_kwarg=1):
+                return do_stuff(uuid_arg, string_arg, kwarg)
+
+        The first call is cached and reused for 120s. If the defaults are inadequate use:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM", "desired_cache_key_name", 120)
+            def my_other_function...
+
+    Args:
+        pool: A async redis cache pool.
+        secret: used to generate a checksum too ensure cache tampering is not possible.
+        prefix: Prefix for the cache keys generated.
+        key_name: When specified the total redis key_name will be "prefix:key_name".
+        expiry: expiration in seconds. Defaults to two minutes (120s).
+
+    Returns:
+        decorator function
+
+    """
+
+    def cache_decorator(func: Callable) -> Callable:
+
+        # Todo: write a sync implementation?
+
+        @wraps(func)
+        async def func_wrapper(*args, **kwargs):
+            if key_name:
+                cache_key = f"{prefix}:{key_name}"
+            else:
+                # Auto generate a cache key name based on function_name and a hash of the arguments
+                # Note: this makes no attempt to handle non-hashable values like lists and sets or other complex objects
+                # Todo: log/warning about key problems for unsupported kwargs argument types?
+                kwd_mark = object()  # sentinel for separating args from kwargs
+                cache_key = f"{prefix}:" + args + (kwd_mark,) + tuple(sorted(kwargs.items()))
```

Comment:
The key can be namespaced to prevent collisions with other stuff. Also, including the python version in the key will safeguard from pickling errors when updating (or switching) to another python version. Maybe something like this;


```python
namespace = "cached_result:{}.{}".format(*sys.version_info[:2])   # "cached_result:3.10"
cache_key = f"{namespace}:{prefix}:{key_name}"
```

Reply:
hmm: the prefix is mandatory -> and should be enough for "namespacing"

I think I like the approach to pickle and python versions. @mrijk Opted to put a try/except around the pickle; and re-cache the value in the except clause. Which would "heal" the cache after a python update.

I think I like the option that uses python version in the key name more: as it would also support running a couple of workers in python3.10 and the rest of the backend in 3.11

Reply:
I think both are needed: e.g. when load from Pickle fails -> reset that cache key